src/alarm_heap.cpp

```diff
@@ -0,0 +1,326 @@
+  _terminated(false),
+  _alarm_table_defs(alarm_table_defs),
+  _alarm_trap_sender(new AlarmTrapSender(this))
+{
+  // Create the lock/condition variables.
+  pthread_mutex_init(&_lock, NULL);
+  _cond = new CondVar(&_lock);
+
+  // Populate the alarm state map
+  for (AlarmTableDefsIterator it = _alarm_table_defs->begin();
+                              it != _alarm_table_defs->end();
+                              it++)
+  {
+    AllAlarmsStateMap::iterator index = _all_alarms_state.find(it->alarm_index());
+
+    if (index == _all_alarms_state.end())
+    {
+      // We only have one entry for each alarm index (not one per severity).
+      AlarmInHeap* alarm_in_heap = new AlarmInHeap(it->alarm_index());
+      AlarmInfo* alarm_info = new AlarmInfo(alarm_in_heap,
+                                            AlarmDef::Severity::UNDEFINED_SEVERITY);
+      _all_alarms_state.insert(std::pair<unsigned int, AlarmInfo*>(it->alarm_index(),
+                                                                   alarm_info));
+    }
+  }
+
+  // Finally, create the heap thread. This covers getting any alarms to send
+  // from the heap and passing them to the trap sender to actually send.
+  int rc = pthread_create(&_heap_sender_thread, NULL, heap_sender_function, this);
+
+  if (rc < 0)
+  {
+    // LCOV_EXCL_START
+    printf("Failed to start heap pop thread: %s", strerror(errno));
+    exit(2);
+    // LCOV_EXCL_STOP
+  }
+}
+
+AlarmHeap::~AlarmHeap()
+{
+  pthread_mutex_lock(&_lock);
+  _terminated = true;
+  _cond->signal();
+  pthread_mutex_unlock(&_lock);
+
+  pthread_join(_heap_sender_thread, NULL);
+  delete _cond; _cond = NULL;
+
+  pthread_mutex_destroy(&_lock);
+
+  delete _alarm_trap_sender; _alarm_trap_sender = NULL;
+  _alarm_heap.clear();
+
+  for (AllAlarmsStateMap::iterator it = _all_alarms_state.begin();
+                               it != _all_alarms_state.end();
+                               it++)
+  {
+    delete (*it).second;
+  }
+}
+
+void* AlarmHeap::heap_sender_function(void* data)
+{
+  ((AlarmHeap*)data)->heap_sender();
+  return NULL;
+}
+
+void AlarmHeap::heap_sender()
+{
+  pthread_mutex_lock(&_lock);
+
+  while (!_terminated)
+  {
+    uint64_t time_now_in_ms = Utils::get_time();
+    AlarmInHeap* alarm_in_heap = (AlarmInHeap*)_alarm_heap.get_next_timer();
+
+    while ((alarm_in_heap) &&
+           (alarm_in_heap->get_pop_time() <= time_now_in_ms))
+    {
+      // While the next alarm in the heap is due to be sent, we:
+      //  - Pull out the alarm definition for its index and severity.
+      //  - Pass the alarm definition to the trap sender (which is responsible
+      //    for actually sending any INFORMs).
+      //  - Update the master view of the current severity for this alarm (by
+      //    updating the _all_alarms_state map).
+      //  - Finally, we then remove the alarm from the heap.
+      AlarmTableDef& alarm_table_def =
+              _alarm_table_defs->get_definition(alarm_in_heap->index(),
```

Comment:
Should the alarm timer just store all this context so you don't have to do these lookups?